src/libos/src/fs/sefs/sgx_storage.rs

```diff
@@ -180,64 +201,109 @@ unsafe impl Sync for LockedFile {}
 
         let offset = offset as u64;
-        file.seek(SeekFrom::Start(offset))
-            .expect("failed to seek SgxFile");
-        let len = file.read(buf).expect("failed to read SgxFile");
+        try_sgx_file_io!(file.seek(SeekFrom::Start(offset)))?;
+        let len = try_sgx_file_io!(file.read(buf))?;
         Ok(len)
     }
 
-    fn write_at(&self, buf: &[u8], offset: usize) -> DevResult<usize> {
+    fn _write_at(&self, buf: &[u8], offset: usize) -> SgxDevResult<usize> {
         if buf.len() == 0 {
             return Ok(0);
         }
         let mut file = self.0.lock().unwrap();
 
         // SgxFile does not support to seek a position beyond the end.
         // So check if file_size < offset and padding null bytes.
-        let file_size = file.seek(SeekFrom::End(0)).expect("failed to tell SgxFile") as usize;
+        let file_size = try_sgx_file_io!(file.seek(SeekFrom::End(0)))? as usize;
         if file_size < offset {
             static ZEROS: [u8; 0x1000] = [0; 0x1000];
             let mut rest_len = offset - file_size;
             while rest_len != 0 {
                 let l = rest_len.min(0x1000);
-                let len = file.write(&ZEROS[..l]).expect("failed to write SgxFile");
+                let len = try_sgx_file_io!(file.write(&ZEROS[..l]))?;
                 rest_len -= len;
             }
         }
 
         let offset = offset as u64;
-        file.seek(SeekFrom::Start(offset))
-            .expect("failed to seek SgxFile");
-        let len = file.write(buf).expect("failed to write SgxFile");
+        try_sgx_file_io!(file.seek(SeekFrom::Start(offset)))?;
+        let len = try_sgx_file_io!(file.write(buf))?;
         Ok(len)
     }
 
-    fn set_len(&self, len: usize) -> DevResult<()> {
+    fn _set_len(&self, len: usize) -> SgxDevResult<()> {
         // NOTE: do nothing ??
         Ok(())
     }
 
-    fn flush(&self) -> DevResult<()> {
+    fn _flush(&self) -> SgxDevResult<()> {
         let mut file = self.0.lock().unwrap();
-        file.flush().expect("failed to flush SgxFile");
+        try_sgx_file_io!(file.flush())?;
         Ok(())
     }
 
-    fn get_file_mac(&self) -> DevResult<SefsMac> {
+    fn _get_file_mac(&self) -> SgxDevResult<SefsMac> {
         let file = self.0.lock().unwrap();
         Ok(SefsMac(file.get_mac().unwrap()))
     }
 }
+
+/// An error type for SGX protected file I/O
+pub struct SgxDevError(i32);
+
+/// A result type for SGX protected file I/O
+pub type SgxDevResult<T> = std::result::Result<T, SgxDevError>;
+
+impl From<std::io::Error> for SgxDevError {
+    fn from(e: std::io::Error) -> Self {
+        if let Some(os_err) = e.raw_os_error() {
+            SgxDevError(os_err)
+        } else if let Some(sgx_err) = e.raw_sgx_error() {
+            SgxDevError(sgx_err as i32)
+        } else {
+            warn!("unknown io error: {:?}, set SgxDevError to EIO", e);
+            SgxDevError(EIO as i32)
+        }
+    }
+}
+
+impl ToDevError for SgxDevError {
```

Comment:
This comment is related to the previous comment about the necessity of introducing `SgxDevError`. 

I think one reason of introducing `SgxDevError` is that you need a type to implement this trait. But this trait is trivial: `Into<DevError>` can serve the same purpose without a new type. Plus, `SgxDevError` is a bit of confusing since we already have a `DevError` type from SEFS. `impl Into<DevError> for Error` should have made it feasible to reuse the `Error` type here, rather than introducing `SgxDevError`.